src/validation.cpp

```diff
@@ -1281,6 +1315,130 @@ static void CheckForkWarningConditionsOnNewFork(CBlockIndex* pindexNewForkTip) E
     CheckForkWarningConditions();
 }
 
+/**
+ * Called when a header (re-)reached BLOCK_VALID_TREE.
+ *
+ * setBlockIndexHeaderCandidates is a bit more complicated than
+ * setBlockIndexCandidates as setBlockIndexCandidates can be rather lazy
+ * as everything in it is about to be connected. In our case, we may have many
+ * headers above the tip leading down different chains, for which we really
+ * only want to keep the tip of each chain.
+ *
+ * Works even if chainActive is empty!
+ *
+ * If chain_ordered_inertion, we assume that if pindex->pprev was previously a
+ * header candidate, it will be when we're called. ie we assume that there are
+ * no header candidates which are parents of us except for possibly our direct
+ * parent.
+ *
+ * no_descendants allows us to make a similar, but inverted, assumption -
+ * assuming no descendant blocks may be header candidates.
+ */
+void CChainState::MaybeAddNewHeaderCandidate(CBlockIndex* pindex, bool chain_ordered_insertion, bool no_descendants) {
+    if (!pindex->IsValid(BLOCK_VALID_TREE)) return; // We only want things that have a valid header tree
+
+    bool lower_work = chainActive.Tip() != nullptr && chainActive.Tip()->nChainWork > pindex->nChainWork;
+    if (lower_work) return; // We don't want things with less work than our current tip
+
+    bool parent_present = false;
+    if (pindex->pprev && setBlockIndexHeaderCandidates.count(pindex->pprev)) {
+        // If the parent is a previous candidate, then no parents of it could
+        // be candidates, either. This is the only thing we need to check by
+        // definition of chain_ordered_insertion, however even in the case of
+        // !chain_ordered_insertion, if this is true, no need to do a full
+        // parent scan (as no further-up parent can be a candidate, either).
+        setBlockIndexHeaderCandidates.erase(pindex->pprev);
+        parent_present = true;
+    } else if (!chain_ordered_insertion) {
+        // We are being called in a for(p : mapBlockIndex) loop, so can make no
+        // assumptions about existing entries. Scan all other entries to check
+        // if we're a descendant of some other candidate.
+        for (auto it = setBlockIndexHeaderCandidates.begin(); it != setBlockIndexHeaderCandidates.end() && (*it)->nChainWork < pindex->nChainWork; it++) {
+            if (pindex->GetAncestor((*it)->nHeight) == *it) {
+                // it should be removed - we only keep the tip of potential
+                // chains, not anything in them. At this point we should be
+                // consistent by adding pindex, there should be more more work
+                // to do here.
+                setBlockIndexHeaderCandidates.erase(it);
+                break;
+                parent_present = true;
```

Comment:
This is dead code. The `parent_present = true;` is not meant to be after `break;` in this scope? :-)